presto-main/src/main/java/com/facebook/presto/operator/aggregation/histogram/ValueStore.java

```diff
@@ -0,0 +1,187 @@
+    private static final GroupedTypedHistogram.ProbeType PROBE_TYPE = SUM_OF_SQUARE;
+    private static final int EMPTY_BUCKET = -1;
+    private final BlockBuilder values;
+    private int rehashCount;
+
+    private int mask;
+    private int bucketCount;
+    private IntBigArray buckets;
+    private IntBigArray valueHashList;
+    private int maxfill;
+
+    @VisibleForTesting
+    public ValueStore(int expectedSize, BlockBuilder values)
+    {
+        bucketCount = computeBucketCount(expectedSize, MAX_FILL_RATIO);
+        mask = bucketCount - 1;
+        maxfill = calculateMaxFill(bucketCount, MAX_FILL_RATIO);
+        this.values = values;
+        buckets = new IntBigArray(-1);
+        buckets.ensureCapacity(bucketCount);
+        valueHashList = new IntBigArray(-1);
+        valueHashList.ensureCapacity(bucketCount);
+    }
+
+    /**
+     * cryptically named, this will add an item if not already in the system. It returns a pointer that is unique for multiple instances of the value
+     *
+     * @param type
+     * @param block
+     * @param position
+     * @param valueOnlyRawHash
+     * @return
+     */
+    public int getPosition(Type type, Block block, int position, long valueOnlyRawHash)
+    {
+        if (values.getPositionCount() >= maxfill) {
+            rehash();
+        }
+
+        long valueHash = valueOnlyRawHash;
+        int bucketId = getHashPosition(valueHash, mask);
+        int valuePointer;
+
+        // look for an empty slot or a slot containing this key
+        int probeCount = 1;
+        int originalBucketId = bucketId;
+        while (true) {
+            valuePointer = buckets.get(bucketId);
+            // empty
+            if (valuePointer == EMPTY_BUCKET) {
+                break;
+            }
+            // value match
+            if (type.equalTo(block, position, values, valuePointer)) {
+                break;
+            }
+            int probe = nextProbe(probeCount);
+            bucketId = nextBucketId(originalBucketId, mask, probe);
+            probeCount++;
+        }
+
+        if (valuePointer == EMPTY_BUCKET) {
+            valuePointer = values.getPositionCount();
+            type.appendTo(block, position, values);
+            buckets.set(bucketId, valuePointer);
+        }
+
+        checkState(probeCount < bucketCount, "we can't search more probe %s more than buckets %d", probeCount, bucketCount);
+
+        valueHashList.set(valuePointer, (int) valueHash);
+
+        return valuePointer;
+    }
+
+    private int getHashPosition(long valueHash, int mask)
+    {
+        return (int) (murmurHash3(valueHash) & mask);
+    }
+
+    @VisibleForTesting
+    void rehash()
+    {
+        ++rehashCount;
+
+        long newBucketCountLong = bucketCount * 2L;
+
+        if (newBucketCountLong > Integer.MAX_VALUE) {
+            throw new PrestoException(GENERIC_INSUFFICIENT_RESOURCES, "Size of hash table cannot exceed 1 billion entries");
```

Comment:
i don't either
it's from the original code, and i spent my time understnading and working on other parts

i guess you could just double bucketCount...


yea, it's 2B I guess. you want me to + the intMax?

Reply:
i think the idea with the local var was to update all state vars at the end (but it was from the original). so if anything failed in the rehash, we had a valid hash table to debug

is it just the *Long version you don't like?

or that we have a newBucketCount?

Reply:
and you need a long version to start with to check intMax, or else you never know how overflow will manifest (2-comp should go negative, but still)

how would you do it?

1. assume you need a long to check for overflow
2. you need the variable as an int several times after that, so it logically makes sense to have an uncasted (int) version to check, then cast down for the rest. 

maybe quick chat on messenger to resolve tmr, or talk in person

i'm leaving as-is, because removing it gets into 
a. doing the *2L twice (or more)
b. casting many times to assignment to ints

Reply:
fixed to use Integer.MAX_VALUE